evm/vm/base.py

```diff
@@ -123,29 +122,63 @@ def execute_bytecode(self,
         )
 
     @abstractmethod
-    def make_receipt(self, transaction, computation, state):
+    def make_receipt(self, base_header, transaction, computation, state):
         """
-        Make receipt.
+        Generate the receipt resulting from applying the transaction.
+
+        :param base_header: the header of the block before the transaction was applied.
+        :param transaction: the transaction used to generate the receipt
+        :param computation: the result of running the transaction computation
+        :param state: the resulting state, after executing the computation
+
+        :return: receipt
+        """
+        raise NotImplementedError("Must be implemented by subclasses")
+
+    @abstractmethod
+    def validate_transaction_against_header(self, base_header, transaction):
+        """
+        Validate that the given transaction is valid to apply to the given header.
+
+        :param base_header: header before applying the transaction
+        :param transaction: the transaction to validate
+
+        :raises: ValidationError if the transaction is not valid to apply
         """
         raise NotImplementedError("Must be implemented by subclasses")
 
-    def apply_transaction(self, transaction):
+    def apply_transaction(self, header, transaction):
         """
         Apply the transaction to the vm in the current block.
+
+        :param transaction: to apply
+        :param header: header of the block before application
         """
+        self.validate_transaction_against_header(header, transaction)
         state_root, computation = self.state.apply_transaction(transaction)
-        receipt = self.make_receipt(transaction, computation, self.state)
-        # TODO: remove this mutation.
-        self.state.gas_used = receipt.gas_used
+        receipt = self.make_receipt(header, transaction, computation, self.state)
 
-        new_header = self.block.header.copy(
-            bloom=int(BloomFilter(self.block.header.bloom) | receipt.bloom),
+        new_header = header.copy(
+            bloom=int(BloomFilter(header.bloom) | receipt.bloom),
             gas_used=receipt.gas_used,
             state_root=state_root,
         )
 
         return new_header, receipt, computation
 
+    def _apply_all_transactions(self, transactions, base_header):
+        receipts = []
+        previous_header = base_header
+        result_header = base_header
+
+        for transaction in transactions:
+            result_header, receipt, _ = self.apply_transaction(previous_header, transaction)
+
+            previous_header = result_header
+            receipts.append(receipt)
+
+        return result_header, receipts
```

Comment:
It was nice to pull this chunk of logic out, but it was awkward to both:

- chain the result of one application into the next, and
- accumulate the results

A small version involved a lot of variable clobbering, like:

```py
    def _apply_all_transactions(self, transactions, header):
        receipts = []

        for transaction in transactions:
            header, receipt, _ = self.apply_transaction(header, transaction)
            receipts.append(receipt)

        return header, receipts
```

I'm happy to hear suggestions about a way to do chaining & accumulation at the same time, in a more elegant way.

Reply:
Thanks! It got me thinking of some other ways to look at it.

Right now, to my eyes, that particular implementation is a drop in readability, compared to having some variable reuse in a short function.

I have an idea for an improvement to get rid of it, but I don't want to throw this PR off track, so I'll put it in the next one.

Reply:
I think in implementation it will look like this.

```python
# top level helper function.
def _accumulate_apply_txn_fn(prev_result, apply_transaction_fn):
    return apply_transaction_fn(prev_result[0])

def import_block(self, block):
    ...
    if block.transactions:
        apply_transaction_fns = tuple(
            functools.partial(self.apply_transaction, transaction=transaction)
            for transaction in block.transactions
        )
        headers, receipts = zip(*tuple(accumulate(
            _accumulate_apply_txn_fn, 
            apply_transaction_fns,
            (self.block.header, None),
        ))[1:])
        last_header = headers[-1]
    else:
        last_header = self.block.header
        receipts = tuple()
     ...
```

I think I generally agree about the readability, but I'm also inclined to cut the extra API from the `VM` class.  We could still use the logic above to encapulate the act of applying all of the transactions and getting back the resulting header and list of receipts but still keep it encapsulated in a utility function to keep the `import_block` method easier to read.

Dunno, your call.  Just wanted to see it all written out mostly.